### ISSUE TITLE
Cursor position is incorrect with markup based editing

### ISSUE THREAD

**Issue by Udayendu** [uiwjs/react-md-editor#541]
When trying to write the page with markup, cursor position is not mapped correctly:

![image](https://github.com/uiwjs/react-md-editor/assets/1809177/ecc3e8ae-2cad-4431-b234-d78fb6620e54)

As you can see above, After a h4 type heading when I am trying to start writing the heading title, the cursor is still two positions back. But the actually it should look something like `#### |`. For normal text its not an issue but specially with heading type content, its causing the visualisation issue.

how this can be fixed ?

**Comment by jaywcjlove** [comment]
<img width="641" alt="image" src="https://github.com/uiwjs/react-md-editor/assets/1680273/17af3bbc-dcdc-44c6-8810-32330ac139a7">

@Udayendu I cannot reproduce your error.

**Comment by Udayendu** [comment]
In my case every time its reproducible. I am not using any CSS for the textarea. Here is the code and the CSS its detecting:

```
<MDEditor
            height={600}
            value={content}
            onChange={handleContent}
            preview="edit"
          />
```

![image](https://github.com/uiwjs/react-md-editor/assets/1809177/2197d93e-c614-45e6-8fd2-d816b7fcba19)

And font is 'Roboto, sans-serif'. So not sure whats impacting it.

**Comment by jaywcjlove** [comment]
@Udayendu Can you provide an example where the error can be reproduced? 


like this: https://codesandbox.io/embed/markdown-editor-for-react-izdd6?fontsize=14&hidenavigation=1&theme=dark

**Comment by Udayendu** [comment]
@jaywcjlove Seems like its a white space issue:

![image](https://github.com/uiwjs/react-md-editor/assets/1809177/9fc5c9bc-7b06-4625-92d4-4c4f4b1085d9)

Take a look into the above image. The actual text is visible after selecting but the one we can see in the MarkDown editor is without white-space. That's why its not able to handle the cursor properly and positioned two spaces back.

I tried editing the style property, but still no improvement:

```
<MDEditor
      height={600}
      value={content}
      onChange={handleContent}
      preview="live"
      style={{ whiteSpace: "pre-wrap" }}
    />
```

Seems like the style properly have no impact in the MDEditor.

**Comment by matt2112** [comment]
I have the same problem. In this screenshot the cursor is actually at the end of the line, but clearly doesn't look that way:

![image](https://github.com/uiwjs/react-md-editor/assets/12831300/5f6e0f33-b4f7-44ec-8058-fe545d641696)

Tried upgrading to latest versions (`@uiw/react-md-editor`: 3.23.3, `@uiw/react-markdown-preview`: 4.1.13) but didn't help.

And here's how I'm using it (with Next 13.4.1):

```ts
import '@uiw/react-markdown-preview/markdown.css'
import '@uiw/react-md-editor/markdown-editor.css'

export const DisplayRow = styled.div`
  display: flex;
  align-items: flex-start;
  padding: ${({ theme }) => theme.spacing(2)} 0;
`

export const DisplayLabel = styled(EmphasisedLabel)`
  color: ${({ theme }) => theme.colors.textEmphasis};
`

export const DisplayInputContainer = styled.div`
  flex: 1;
`

const MDEditor = dynamic(() => import('@uiw/react-md-editor'), { ssr: false })

const MarkdownField = ({ label, value, setValue }: MarkdownFieldProps) => (
  <DisplayRow>
    <DisplayLabel>{label}:</DisplayLabel>
    <DisplayInputContainer>
      <MDEditor
        aria-label={label}
        value={value}
        onChange={(val) => setValue(val || '')}
        height={120}
        data-color-mode="light"
        className={label}
      />
    </DisplayInputContainer>
  </DisplayRow>
)
```

Hope that's helpful. Great library though, appreciate the work that goes into it!

**Comment by matt2112** [comment]
Hmm so yeah I converted the above code to a Codesandbox and you're right @jaywcjlove that there's no issue 🤔 https://codesandbox.io/p/sandbox/exciting-pine-72x535

@Udayendu I don't suppose you're also using Ant Design? I'm now wondering if there's some kind of style conflict with that somewhere.

**Comment by matt2112** [comment]
Ah so I think I see the issue. It appears it's because of this line:

https://github.com/uiwjs/react-md-editor/blob/d7ba73b15de9faf8811fa9a3ea0f7c69b80531cb/core/src/components/TextArea/index.less#L30

i.e. the span has the font family that we use elsewhere in the app (in our case that's `Gilroy, Arial, Helvetica, sans-serif`), but the textarea has `"Helvetica Neue", Helvetica, Arial, sans-serif`. Therefore, the cursor gets slightly out of sync. If I override the font-family on the textarea then it all lines up nicely.

@Udayendu give that a try and let me know if it solves it for you too.

@jaywcjlove is there a good reason that line needs to have the `!important`?

**Comment by matt2112** [comment]
@jaywcjlove Sorry unsure if that's a suggestion or an explanation. I mean why does the library override the application's font family with the !important flag? Surely in most cases developers will want to use the same font as they have everywhere else no? But could well be something I'm missing.

**Comment by jaywcjlove** [comment]
Yes, it will not easily cause style conflicts when users use it.

https://github.com/uiwjs/react-md-editor/blob/d7ba73b15de9faf8811fa9a3ea0f7c69b80531cb/core/src/components/TextArea/index.less#L54

```css
.w-md-editor-text-input,
.w-md-editor-text-pre > code,
.w-md-editor-text-pre 
{
  font-family: Arial,sans-serif !important;
}
```

or 

```css
.w-md-editor {
  --md-editor-font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif !important;
}
```

**Comment by Udayendu** [comment]
> Ah so I think I see the issue. It appears it's because of this line:
> 
> https://github.com/uiwjs/react-md-editor/blob/d7ba73b15de9faf8811fa9a3ea0f7c69b80531cb/core/src/components/TextArea/index.less#L30
> 
> i.e. the span has the font family that we use elsewhere in the app (in our case that's `Gilroy, Arial, Helvetica, sans-serif`), but the textarea has `"Helvetica Neue", Helvetica, Arial, sans-serif`. Therefore, the cursor gets slightly out of sync. If I override the font-family on the textarea then it all lines up nicely.
> 
> @Udayendu give that a try and let me know if it solves it for you too.
> 
> @jaywcjlove is there a good reason that line needs to have the `!important`?

@matt2112  I tried applying this fix but some one its not able to resolve my issue. Are you able to fix your issue ?

**Comment by jaywcjlove** [comment]
![image](https://github.com/uiwjs/react-md-editor/assets/1680273/27b4a3a2-1e02-4955-9bbd-78d2da027f89)

https://codesandbox.io/embed/markdown-editor-for-react-https-github-com-uiwjs-react-md-editor-issues-541-46nflr?fontsize=14&hidenavigation=1&theme=dark

I'm not sure what the problem is.

**Comment by Udayendu** [comment]
@jaywcjlove I tried to apply it using the CSS file and I can still see the issue. As you can see in the below image the font family is getting detected but its not applying the changes.

<img width="1498" alt="Screenshot 2023-06-29 at 6 28 51 PM" src="https://github.com/uiwjs/react-md-editor/assets/1809177/cec01444-31ce-4fef-8088-3f1712b1aa3c">


I tried to check the outcome on multiple browser like firefox, chrome and edge. But still the behaviour is save. And even on both Ubuntu and MacOS too. Not sure why its not getting applied correctly.

I create a test project and use the editor and its working there. But not in the main project.

**Comment by Udayendu** [comment]
And this time I applied directly in the JS file using style component the way you did, but still not working:

<img width="1498" alt="Screenshot 2023-06-29 at 6 38 45 PM" src="https://github.com/uiwjs/react-md-editor/assets/1809177/bf3e1996-8d27-4028-9e1f-eae8755c244a">

I know its not the editor thats breaking but something is there might be in my code which is not working as expected.

**Comment by Udayendu** [comment]
@jaywcjlove  btw FYI, this issue is happening when I am using any kind of markup parameters. For the normal texts, there is no issue atall.

**Comment by Udayendu** [comment]
I just did some more testing and seems like apart from "Helvetica Neue" font, all other fonts have the issue and its not installed in the server thats hosting this application. You can see some dummy code https://codesandbox.io/s/white-dew-z786fl 

But usually these fonts required at the client side, is not it ?

**Comment by jaywcjlove** [comment]
@Udayendu https://codesandbox.io/embed/serene-platform-j4ywzp?fontsize=14&hidenavigation=1&theme=dark

```diff
- .w-md-editor-text-input {
+ .w-md-editor {
  /* --md-editor-font-family: "Helvetica Neue", Helvetica, Arial, sans-serif !important; */
  --md-editor-font-family: Arial;
}
```

**Comment by Udayendu** [comment]
Thanks @jaywcjlove 
After making this change, I can see that majority of the font issues are gone now. 

But when I am trying to make the title in any place of the page, I can see this issue.
Here is what I can see in the console as well.

<img width="1465" alt="Screenshot 2023-07-01 at 8 19 18 AM" src="https://github.com/uiwjs/react-md-editor/assets/1809177/ef1fe064-2327-40c9-bfa8-a144e4d311a1">

Do you think something is trying to override some properly related to font ??

**Comment by jaywcjlove** [comment]
<img width="342" alt="image" src="https://github.com/uiwjs/react-md-editor/assets/1680273/307c3684-9b72-445d-ad3b-c25f11ab62eb">

@Udayendu I can't help you if you give me an example where the error can be reproduced.

**Comment by Udayendu** [comment]
@jaywcjlove  Its difficult to reproduce out side of my project. The issue is that somewhere my code is causing a conflict with the required font for this editor. And may be thats causing the issue. Is there any way to override the default font and use only what we have specified for this editor. I know we have already tried with `!important` but some how thats not helping much. And if I am starting a line with markup, its breaking the design.

So I am really not sure how can I give you a reproducer.

**Comment by Udayendu** [comment]
Closing this issue as I am unable to reproduce it outside of my project.

**Comment by Jadiction** [comment]
I know this is an old issue, but I ran into this issue while searching for a fix and wanted to share my fix in case others run into it.

With enough messing around, I learned it only happened when the line had to wrap. (Issue wasn't replicating when full-screen or shorter text)

At first, I fixed it with a white-space: break spaces; to the text input.

```css
.editor :global(.w-md-editor-text-input) {
  white-space: break-spaces !important;
}
```
but when doing no-space line wraps, the issue persisted.

The final fix for me is:
```css
.quest :global(.w-md-editor-text),
.quest :global(.w-md-editor-text-pre),
.quest :global(.w-md-editor-text-input),
.quest :global(.w-md-editor-text-pre > code) {
  max-width: 100% !important;
  white-space: break-spaces !important;
}
```
In these example, my MDEditor is wrapped with the "editor" div, so you'll have to adjust the selector accordingly.

Example of my issue:
https://i.imgur.com/Idtxg2Y.gif

This is likely caused by CSS confliction, I just didn't do much testing as to what caused it after I fixed it.
